// @magidoc/cli/build/tasks/all/loadGraphqlSchema.js
import { parseSchema, fetchSchema } from '@magidoc/rollup-plugin-gql-schema';
import fs from 'fs/promises';

function loadGraphQLSchemaTask(config) {
    return {
        title: `Load GraphQL Schema`,
        executor: async (ctx) => {
            switch (config.introspection.type) {
                case 'url':
                    await fetchSchema({
                        url: config.introspection.url,
                        method: config.introspection.method,
                        query: config.introspection.query,
                        headers: config.introspection.headers,
                        target: ctx.templateConfiguration.schemaTargetLocation,
                        format: 'sdl',
                    });
                    break;
                case 'sdl':
                    await parseSchema({
                        paths: config.introspection.paths,
                        target: ctx.templateConfiguration.schemaTargetLocation,
                        format: 'sdl',
                    });
                    break;
                case 'raw':
                    await fs.writeFile(ctx.templateConfiguration.schemaTargetLocation, config.introspection.content);
                case 'none':
                    await fs.writeFile(ctx.templateConfiguration.schemaTargetLocation, '');
            }
        },
    };
}

export { loadGraphQLSchemaTask };
